Sources/SourceKit/sourcekitd/SemanticRefactoring.swift

```diff
@@ -0,0 +1,157 @@
+           let text: String = value[keys.text]
+        {
+          // The LSP is zero based, but semantic_refactoring is one based.
+          let startPosition = Position(line: startLine - 1, utf16index: startColumn - 1)
+          let endPosition = Position(line: endLine - 1, utf16index: endColumn - 1)
+          let edit = TextEdit(range: startPosition..<endPosition, newText: text)
+          textEdits.append(edit)
+        }
+        return true
+      }
+      return true
+    }
+
+    guard textEdits.isEmpty == false else {
+      return nil
+    }
+
+    self.title = title
+    self.edit = WorkspaceEdit(changes: [url: textEdits])
+  }
+}
+
+/// An error from a cursor info request.
+enum SemanticRefactoringError: Error {
+
+  /// The given URL is not a known document.
+  case unknownDocument(URL)
+
+  /// The given position range is invalid.
+  case invalidRange(Range<Position>)
+
+  /// The underlying sourcekitd request failed with the given error.
+  case responseError(ResponseError)
+
+  /// The underlying sourcekitd reported no edits for this action.
+  case noEditsNeeded(URL)
+}
+
+extension SemanticRefactoringError: CustomStringConvertible {
+  var description: String {
+    switch self {
+    case .unknownDocument(let url):
+      return "failed to find snapshot for url \(url)"
+    case .invalidRange(let range):
+      return "failed to refactor due to invalid range: \(range)"
+    case .responseError(let error):
+      return "\(error)"
+    case .noEditsNeeded(let url):
+      return "no edits reported for semantic refactoring action for url \(url)"
+    }
+  }
+}
+
+extension SwiftLanguageServer {
+
+  /// Provides detailed information about the result of a specific refactoring operation.
+  ///
+  /// Wraps the information returned by sourcekitd's `semantic_refactoring` request, such as the necessary edits and placeholder locations.
+  ///
+  /// - Parameters:
+  ///   - url: Document URL in which to perform the request. Must be an open document.
+  ///   - command: The semantic refactor `Command` that triggered this request.
+  ///   - completion: Completion block to asynchronously receive the SemanticRefactoring data, or error.
+  func semanticRefactoring(
+    _ refactorCommand: SemanticRefactorCommand,
+    _ completion: @escaping (Result<SemanticRefactoring, SemanticRefactoringError>) -> Void)
+  {
+    let url = refactorCommand.textDocument.url
+    guard let snapshot = documentManager.latestSnapshot(url) else {
+      return completion(.failure(.unknownDocument(url)))
+    }
+    guard let startOffset = snapshot.utf8Offset(of: refactorCommand.positionRange.lowerBound),
+          let endOffset = snapshot.utf8Offset(of: refactorCommand.positionRange.upperBound) else
+    {
+      return completion(.failure(.invalidRange(refactorCommand.positionRange)))
+    }
+    let skreq = SKRequestDictionary(sourcekitd: sourcekitd)
+    skreq[keys.request] = requests.semantic_refactoring
+    skreq[keys.name] = ""
```

Comment:
Worth a comment about what this is: 
// Preferred name for e.g. an extracted variable. Empty string means sourcekitd chooses a name automatically.